### PR TITLE
Representing type hierarchies

### PR DESCRIPTION
**Pull request by fsteeg**
Current work in progress for #68.

### CODE REVIEW

**Comment by fsteeg** [comment]
As a basic test case for this, I've added single-level `broader` type support to our reconciliation service at https://test.lobid.org/gnd/reconcile (see https://github.com/hbz/lobid-gnd/commit/5c8d0a8a2049b06dee23827e6e33cf20065b3f1f) and to the testbench (see https://github.com/reconciliation-api/testbench/pull/28). I think this shows how even a simple implementation of `broader` IDs can be useful. In the future, we could look into full hierarchy support in the testbench, if that makes sense.

**Review by wetneb — APPROVED** [review]
Amazing! Nice that you managed to update the testbench and your endpoint in the same go. It looks great to me, but maybe we can wait for our call later today to merge this, in case others have more feedback.

**Comment by orothad** [comment]
```json
{
  "$schema": "http://json-schema.org/schema#",
  "$id": "https://reconciliation-api.github.io/specs/latest/schemas/suggest-types-response.json",
  "type": "object",
  "description": "This schema can be used to validate the JSON response of a suggest service for types.",
  "properties": {
    "result": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "id": {
            "type": "string",
            "description": "Identifier of the suggested type"
          },
          "name": {
            "type": "string",
            "description": "Name of the suggested type"
          },
          "description": {
            "type": "string",
            "description": "An optional description which can be provided to disambiguate namesakes, providing more context."
          },
          "broader": {
            "type": "string",
            "description": "A broader semantic relation as defined by SKOS http://www.w3.org/2004/02/skos/core#broader",
            "$id": "http://www.w3.org/2004/02/skos/core#broader",
            "additionalProperties": {
              "label": "skos:broadMatch"
            }
          }
        },
```

Some clients might want to enrich the presentation of properties with additional metadata or annotations.
In the above example "broader" is given more context and where that context can be displayed by a smart client.

I read though some of the JSON Schema core and appendix for Working with vocabularies, but not 100% sure my example is conforming for validation, but still wanted to show what I was asking about on our last meeting, that of giving more context for a property such as "broader".  And it seems JSON Schema has various ways that more context, annotation, or metadata could be given depending on validation needs, or ignoring validation (the validator ignores, but smart clients can still understand and use)

1. I "think" in the above case that "label" would be ignored by some clients? - https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.10.3.2.3

2. I'm not sure about "$id" but it seems to be valid for JSON Schema? - https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.2.1

**Comment by thadguidry** [comment]
```json
{
  "$schema": "http://json-schema.org/schema#",
  "$id": "https://reconciliation-api.github.io/specs/latest/schemas/suggest-types-response.json",
  "type": "object",
  "description": "This schema can be used to validate the JSON response of a suggest service for types.",
  "properties": {
    "result": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "id": {
            "type": "string",
            "description": "Identifier of the suggested type"
          },
          "name": {
            "type": "string",
            "description": "Name of the suggested type"
          },
          "description": {
            "type": "string",
            "description": "An optional description which can be provided to disambiguate namesakes, providing more context."
          },
          "broader": {
            "type": "string",
            "description": "A broader semantic relation as defined by SKOS http://www.w3.org/2004/02/skos/core#broader",
            "$id": "http://www.w3.org/2004/02/skos/core#broader",
            "additionalProperties": {
              "label": "skos:broadMatch"
            }
          }
        },
```

Some clients might want to enrich the presentation of properties with additional metadata or annotations.
In the above example "broader" is given more context and where that context can be displayed by a smart client.

I read though some of the JSON Schema core and appendix for Working with vocabularies, but not 100% sure my example is conforming for validation, but still wanted to show what I was asking about on our last meeting, that of giving more context for a property such as "broader".  And it seems JSON Schema has various ways that more context, annotation, or metadata could be given depending on validation needs, or ignoring validation (the validator ignores, but smart clients can still understand and use)

1. I "think" in the above case that "label" would be ignored by some clients? - https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.10.3.2.3

2. I'm not sure about "$id" but it seems to be valid for JSON Schema? - https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.2.1

**Comment by fsteeg** [comment]
Switched from array of strings to array of types in https://github.com/reconciliation-api/specs/pull/73/commits/f19ce913afe4e9efadf06e42ada07ca321ceac20 as discussed in https://github.com/reconciliation-api/testbench/pull/28 and https://etherpad.lobid.org/p/reconciliation-november-2021.

Next: update our service implementation in https://github.com/hbz/lobid-gnd/pull/293 and testbench in https://github.com/reconciliation-api/testbench/pull/28.

**Comment by fsteeg** [comment]
> Next: update our service implementation [...] and testbench [...].

Done, see https://github.com/hbz/lobid-gnd/pull/293#issuecomment-993368743 and https://github.com/reconciliation-api/testbench/pull/28#issuecomment-993370187.

File: test.sh [hunk 1]
```diff
@@ -9,8 +9,8 @@ for version in "0.1" "latest"; do
 	for schema_file in `ls $version/schemas`; do
 		echo "Testing $schema_file"
 		schema_name=$(echo "$schema_file" | cut -f 1 -d '.')
-		ajv test -s $version/schemas/$schema_file -d "$version/examples/$schema_name/valid/*.json" --valid
-		ajv test -s $version/schemas/$schema_file -d "$version/examples/$schema_name/invalid/*.json" --invalid
+		ajv test -s $version/schemas/$schema_file -r latest/schemas/type.json -d "$version/examples/$schema_name/valid/*.json" --valid
+		ajv test -s $version/schemas/$schema_file -r latest/schemas/type.json -d "$version/examples/$schema_name/invalid/*.json" --invalid
```
Comment (wetneb) [inline test.sh:13]: I am wondering if there would be a more flexible way to do this… I assume you tried and we can just go with this for now, and think of a better way once this becomes a problem.

**Review by wetneb — APPROVED** [review]

**Pull request merged**